pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -358,6 +369,32 @@ public boolean commitSegment(String rawTableName, final String committingSegment
     oldSegMetadata.setEndTime(now);
     oldSegMetadata.setDownloadUrl(
         ControllerConf.constructDownloadUrl(rawTableName, committingSegmentNameStr, _controllerConf.generateVipUrl()));
+    // Pull segment metadata from incoming segment and set it in zk segment metadata
+    final String baseDir = StringUtil.join("/", _controllerConf.getDataDir(), rawTableName);
+    final String segFileName = StringUtil.join("/", baseDir, committingSegmentNameStr);
+    final File segFile = new File(segFileName);
+    // File must exist since this method is called only after the file is downloaded.
+    SegmentMetadataImpl segmentMetadata;
+    Path metadataPath = null;
+    try {
+      InputStream is = TarGzCompressionUtils
+          .unTarOneFile(new FileInputStream(segFile), V1Constants.MetadataKeys.METADATA_FILE_NAME);
+      metadataPath = FileSystems.getDefault().getPath(baseDir, committingSegmentNameStr + ".metadata");
+      Files.copy(is, metadataPath);
+      segmentMetadata = new SegmentMetadataImpl(new File(metadataPath.toString()));
+    } catch (Exception e) {
+      throw new RuntimeException("Exception extacting and reading segment metadata for " + committingSegmentNameStr, e);
+    } finally {
+      if (metadataPath != null) {
+        FileUtils.deleteQuietly(new File(metadataPath.toString()));
+      }
+    }
+    oldSegMetadata.setCrc(Long.valueOf(segmentMetadata.getCrc()));
+    oldSegMetadata.setStartTime(segmentMetadata.getTimeInterval().getStartMillis());
+    oldSegMetadata.setEndTime(segmentMetadata.getTimeInterval().getEndMillis());
+    oldSegMetadata.setIndexVersion(segmentMetadata.getVersion());
+    oldSegMetadata.setTotalRawDocs(segmentMetadata.getTotalRawDocs());
+
     final ZNRecord oldZnRecord = oldSegMetadata.toZNRecord();
     final String oldZnodePath = PinotRealtimeSegmentManager.getSegmentsPath() + "/" + realtimeTableName + "/" + committingSegmentNameStr;
```

Comment:
Would be good to use utility methods for those, I'm pretty sure there's a method that builds a property store znode path given a table and segment name.